src/storage-client/src/types/connections.rs

```diff
@@ -1029,3 +1056,200 @@ impl SshTunnel {
+    secrets_reader: Arc<dyn SecretsReader>,
+    error: Mutex<Option<String>>,
+    runtime: Handle,
+}
+
+impl<C> DynamicBrokerRewritingClientContext<C> {
+    /// Constructs a new context that wraps `inner`.
+    pub fn new(
+        inner: C,
+        secrets_reader: Arc<dyn SecretsReader>,
+        runtime: Handle,
+    ) -> DynamicBrokerRewritingClientContext<C> {
+        DynamicBrokerRewritingClientContext {
+            inner: BrokerRewritingClientContext::new(inner),
+            default_override: None,
+            secrets_reader,
+            error: Mutex::new(None),
+            runtime,
+        }
+    }
+
+    /// Adds the default broker rewrite rule.
+    ///
+    /// Connections to brokers that aren't specified in other rewrites will be rewritten to connect to
+    /// `rewrite_host` and `rewrite_port` instead.
+    pub fn add_default_broker_rewrite(&mut self, tunnel: SshTunnel) {
+        self.default_override = Some(tunnel);
+    }
+
+    /// Returns a reference to the wrapped context.
+    pub fn inner(&self) -> &C {
+        self.inner.inner()
+    }
+
+    /// Returns any errors that occurred when setting up
+    /// ssh tunnels.
+    pub fn dynamic_error(&self) -> Option<String> {
+        self.error.lock().expect("poisoned").clone()
+    }
+}
+
+impl<C> ClientContext for DynamicBrokerRewritingClientContext<C>
+where
+    C: ClientContext,
+{
+    const ENABLE_REFRESH_OAUTH_TOKEN: bool = C::ENABLE_REFRESH_OAUTH_TOKEN;
+
+    fn rewrite_broker_addr(&self, addr: BrokerAddr) -> BrokerAddr {
+        if self.inner.overridden(&addr) {
+            return self.inner.rewrite_broker_addr(addr);
+        }
+
+        match &self.default_override {
+            Some(default_tunnel) => {
+                let broker = addr.clone();
+                let default_tunnel = default_tunnel.clone();
+                let sr = Arc::clone(&self.secrets_reader);
+                let runtime = self.runtime.clone();
+                // We spawn an additional thread here, and immediately
+                // join it so that the rdkafka thread calling this callback
+                // spends all its time parked, instead of being
+                // continuously woken up when we `block_on` the
+                // ssh tunnel creation.
+                let res = match std::thread::spawn(move || {
+                    runtime.block_on(async move {
+                        let ret = default_tunnel
+                            .build_ssh_tunnel_for_url(
+                                &*sr,
+                                &broker.host,
+                                broker.port.parse().unwrap(),
+                                "kafka default",
+                            )
+                            .await;
+                        ret
+                    })
+                })
+                .join()
+                {
+                    Ok(res) => res,
+                    Err(e) => std::panic::resume_unwind(e),
```

Comment:
`resume_unwind`?

Reply:
this just transfers the panic in the spawned thread into the current thread